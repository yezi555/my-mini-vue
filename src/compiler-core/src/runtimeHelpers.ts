
export const TO_DISPLAY_STRING = Symbol('toDisplayString');

export const CREARE_ELEMENT_VNODE = Symbol('createelementVNode');


export const helperMapName = {
  [TO_DISPLAY_STRING]:"toDisplayString",
  [CREARE_ELEMENT_VNODE]:"createelementVNode"
}