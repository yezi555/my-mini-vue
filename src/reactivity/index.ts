export function reactive(a,b){
  return a + b;
}