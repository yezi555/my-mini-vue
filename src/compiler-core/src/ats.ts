import { CREARE_ELEMENT_VNODE } from "./runtimeHelpers";

export enum NodeTypes{
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION
}

export function createVnodeCall(context:any,tag:any,props:any,children:any){
  context.helper(CREARE_ELEMENT_VNODE);

  return{
    type:NodeTypes.ELEMENT,
    tag,
    props,
    children
  }
}