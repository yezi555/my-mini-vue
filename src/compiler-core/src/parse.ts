import { NodeTypes } from "./ats";

const enum TagTypes {
  Start,
  End,
}

export function baseParse(content:string){

  const  context = createParseCountext(content);

  return createRoot(parseChildren(context,[]))
}


function parseChildren(context:any,ancestors:any){
  const nodes:any = []

  while(!isEnd(context,ancestors)){
    let  node;
    const s = context.source;
    if(s.startsWith("{{")){
      node = parseInterpolation(context)
    }else if(s[0] === '<'){
      if(/[a-z]/i.test(s[1])){
        node =  parseElement(context,ancestors)
      }
    }
    if(!node){
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
}

function isEnd(context:any,ancestors:any){
  //2.遇到结束标签的时候
  const s = context.source;

  if(s.startsWith('</')){
    for (let i = ancestors.length - 1; i >=0; i--) {
      const tag = ancestors[i].tag;
      if(startsWithEndTagOpen(s,tag) ){
        return true
      }
    }
  }

  // if(parentTag && s.startsWith(`</${parentTag}>`)){
  //   return true
  // }
  //1.source有值的时候
  return !s;

}

function parseTextData(context:any,length:number){
  const content = context.source.slice(0, length)
  advanceBy(context, length);
  return content
}

function parseText(context:any){
  let endIndex = context.source.length;
  let endToken = [ "<","{{"];
  for (let i = 0; i < endToken.length; i++) {
    const index = context.source.indexOf(endToken[i]);
    if(index !== -1 && endIndex > index){
      endIndex = index;
    }
  }
  

  const content = parseTextData(context,endIndex);
  return{
    type:NodeTypes.TEXT,
    content
  }
}


function parseElement(context:any,ancestors:any){
  //解析tag
  //删除处理完成的代码
  const element:any =  parseTag(context,TagTypes.Start);
  ancestors.push(element);
  element.children = parseChildren(context,ancestors);
  ancestors.pop();

  if(startsWithEndTagOpen(context.source,element.tag)){
    parseTag(context,TagTypes.End);

  }else{
    throw new Error(`缺少结束标签:${element.tag}`)
  }

  return element;
}

function startsWithEndTagOpen(source:string,tag:string){
  return source.startsWith('<') && source.slice(2,2 + tag.length).toLowerCase() === tag.toLowerCase()
}

function parseTag(context:any,type:TagTypes){
  const match:any = /^<\/?([a-z]*)/i.exec(context.source);
  const tag = match[1];
  advanceBy(context,match[0].length)
  advanceBy(context,1)

  if(type === TagTypes.End) return;
  return{
    type:NodeTypes.ELEMENT,
    tag
  }
}


function parseInterpolation(context:any){
//
  const openDelimiter = "{{";
  const closeDelimiter = "}}";

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  advanceBy(context,openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;
  const rawContent = parseTextData(context,rawContentLength) ;
  const content = rawContent.trim();

  advanceBy(context,closeDelimiter.length);


  return {
    type:NodeTypes.INTERPOLATION,
    content:{
      type:NodeTypes.SIMPLE_EXPRESSION,
      content:content
    }
  }
}

function advanceBy(context:any,length:number){

  context.source =  context.source.slice(length);
}

function createRoot(children:any){
  return {
    children,
  }
}

function createParseCountext(content:string):any{

  return {
    source:content
  }
}