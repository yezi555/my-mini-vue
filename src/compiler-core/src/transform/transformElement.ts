import { NodeTypes } from "../ats";
import { CREARE_ELEMENT_VNODE } from "../runtimeHelpers";


export function transformEelement(node:any,context:any){

  if(node.type === NodeTypes.ELEMENT){

    context.helper(CREARE_ELEMENT_VNODE)
  }
}

