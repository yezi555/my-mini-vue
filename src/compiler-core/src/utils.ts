import { NodeTypes } from "./ats";


export   function isText(node:any){
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION 
}