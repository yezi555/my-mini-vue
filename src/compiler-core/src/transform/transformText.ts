import { NodeTypes } from "../ats";
import { isText } from "../utils";


export function transformText(node:any,context:any){

 
  if(node.type === NodeTypes.ELEMENT){
    return ()=>{
      const {children} = node;
      let currentContainer;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if(isText(child)){
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j];
            if(isText(next)){
              if(!currentContainer){
                currentContainer = children[i] =  {
                  type:NodeTypes.COMPOUND_EXPRESSION,
                  children:[child]
                }
              }
  
              currentContainer.children.push(' + ');
              currentContainer.children.push(next);
              children.slice(j,1);
              j--;
            }else{
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    }
  }
}

