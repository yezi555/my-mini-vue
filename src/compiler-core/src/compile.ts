import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformEelement } from "./transform/transformElement";
import { transformExpression } from "./transform/transformExpression";
import { transformText } from "./transform/transformText";

export function baseCompile(template:any){
  const ast:any = baseParse(template);
  transform(ast,{
    nodeTransforms:[transformExpression,transformEelement,transformText]
  });

  console.log('ast------',ast,ast.codegenNode.children);
  
  return generate(ast);
}