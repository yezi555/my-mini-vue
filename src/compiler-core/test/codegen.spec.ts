import { baseParse } from "../src/parse"
import { generate } from "../src/codegen"
import { transform } from "../src/transform";
import { transformExpression } from "../src/transform/transformExpression";
import { transformEelement } from "../src/transform/transformElement";



describe('codegen',()=>{

    it('string',()=>{
      const ast = baseParse('hi');
      transform(ast);

      const {code} = generate(ast);

      //快照
      //1.抓bug
      //2.有意
      expect(code).toMatchSnapshot();
    });

    it("interpolation",()=>{
      const ast = baseParse('{{message}}');
      transform(ast,{
        nodeTransforms:[transformExpression]
      });

      const {code} = generate(ast);

      //快照
      //1.抓bug
      //2.有意
      expect(code).toMatchSnapshot();
    });
    
    it("element",()=>{
      const ast = baseParse('<div></div>');
      transform(ast,{
        nodeTransforms:[transformEelement]
      });

      const {code} = generate(ast);

      //快照
      //1.抓bug
      //2.有意
      expect(code).toMatchSnapshot();
    })

})