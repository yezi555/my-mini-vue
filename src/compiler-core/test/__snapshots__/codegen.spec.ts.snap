// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `"/nreturnfunction render(_ctx,_cache){return}"`;

exports[`codegen interpolation 1`] = `"const {  } = Vue/nreturnfunction render(_ctx,_cache){return__toDisplayString(_ctx.message)}"`;

exports[`codegen string 1`] = `"/nreturnfunction render(_ctx,_cache){returnhi}"`;
